include/libtorrent/kademlia/dht_storage.hpp

```diff
@@ -0,0 +1,200 @@
+	// constructor function is called dht_default_storage_constructor().
+	//
+	struct TORRENT_EXPORT dht_storage_interface
+	{
+#ifndef TORRENT_NO_DEPRECATE
+		// This function returns the number of torrents tracked by
+		// the DHT at the moment. It's used to fill session_status.
+		// It's deprecated.
+		//
+		virtual size_t num_torrents() const = 0;
+
+		// This function returns the sum of all of peers per torrent
+		// tracker byt the DHT at the moment.
+		// It's deprecated.
+		//
+		virtual size_t num_peers() const = 0;
+#endif
+
+		// This functions retrieve the peers tracked by the DHT
+		// corresponding to the given info_hash. You can specify if
+		// you want only seeds and/or you are scraping the data.
+		//
+		// For future implementers:
+		// If the torrent tracked contains a name, such a name
+		// must be stored as a string in peers["n"]
+		//
+		// If the scrape parameter is true, you should fill these keys:
+		// peers["BFpe"] - with the standard bit representation of a
+		//                 256 bloom filter containing the downloaders
+		// peers["BFsd"] - with the standard bit representation of a
+		//                 256 bloom filter containing the seeders
+		//
+		// If no scraping, you should fill the key peers["values"] with
+		// a list containing a subset of peers tracked by the given
+		// info_hash. Such a list should consider the value of
+		// dht_settings::max_peers_reply. If noseed is true only peers
+		// marked as no seed should be included.
+		//
+		// returns true if an entry with the info_hash is found and
+		// the data is returned inside the (entry) out parameter peers.
+		//
+		virtual bool lookup_peers(sha1_hash const& info_hash
+			, bool noseed, bool scrape
+			, entry& peers) const = 0;
+		// This method is named announce_peer for consistency with the
+		// upper layers, but has nothing to do with networking. Its only
+		// responsibility is store the peer in such a way that it's returned
+		// in the entry with the lookup_peers.
+		//
+		// The name parameter is the name of the torrent if provided during
+		// the announce_peer DHT message. The length of this value should
+		// have a maximum length in the final storage. The default
+		// implementation truncate the value for a maximum of 50 characters.
+		//
+		virtual void announce_peer(sha1_hash const& info_hash
+			, tcp::endpoint const& endp
+			, std::string const& name, bool seed) = 0;
+
+		// This function retrieves the immutable item given the target.
+		//
+		// For future implementers:
+		// The value should be returned as an entry in the key item["v"].
+		//
+		// returns true if the item is found and the data is returned
+		// inside the (entry) out parameter item.
+		//
+		virtual bool get_immutable_item(sha1_hash const& target
+			, entry& item) const = 0;
+		// Store the item's data. This layer is only for storage, the
+		// checking of the target info_hash and the data is performed in
```

Comment:
"info_hash" isn't correct here. The correct description may be a bit involved (since it's the hash of the public key + sequence number + salt and involves the signature). Perhaps it could just say:

"The authentication of the item is performed by the upper layer."